File: solutions/java/0239.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        // Monotonic decreasing deque storing indices of nums, where the leftmost element is the greatest number.
        Deque<Integer> deque = new ArrayDeque<>();
        int[] maxWindow = new int[nums.length - k + 1];
        int l = 0;

        for (int r = 0; r < nums.length; r++) {
            // Remove indices that are out of bounds (outside the current window)
            if (!deque.isEmpty() && deque.peekFirst() < r - k + 1) {
                deque.pollFirst();
            }

            // Remove elements from the deque that are smaller than the current element
            // because they are not useful anymore.
            while (!deque.isEmpty() && nums[deque.peekLast()] < nums[r]) {
                deque.pollLast();
            }

            // Add the current element index to the deque
            deque.addLast(r);

            // Once the window size reaches at least 'k', start adding results
            if (r >= k - 1) {
                maxWindow[l] = nums[deque.peekFirst()];
                l++;
            }
        }
        
        return maxWindow;
    }
}

File: solutions/python/0239.py
"""
Leetcode 239: Sliding Window Maximum. Python. Hard.
O(N) runtime and O(K) space complexity, where N is the number of elements and K is the window size.
Usee a monotonic deque to track max elements in each sliding window.
Remove stale indices as the window moves.
3/31/2025 Winston Tsui
"""

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        maxWindow = []
        deque = collections.deque()

        # Stores indices of elements in the window, front is the max
        l = r = 0
        while r < len(nums):
            # Keep only the greatest element in the back of deque (monotonically decreasing deque)
            while deque and nums[r] > nums[deque[-1]]:
                deque.pop()
            deque.append(r)

            # Remove indicies out of current window
            if deque[0] < l:
                deque.popleft()

            # Append to maxWindow once window reaches size k and update window left boundary
            if r + 1 >= k:
                maxWindow.append(nums[deque[0]])
                l += 1
            r += 1

        return maxWindow
